lib/auth/join_tpm_test.go

```diff
@@ -0,0 +1,344 @@
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/tpm"
+)
+
+type mockTPMValidator struct {
+	lastCalledParams   tpm.ValidateParams
+	returnErr          error
+	returnValidatedTPM *tpm.ValidatedTPM
+}
+
+func (m *mockTPMValidator) setup(returns *tpm.ValidatedTPM, err error) {
+	m.lastCalledParams = tpm.ValidateParams{}
+	m.returnErr = err
+	m.returnValidatedTPM = returns
+}
+
+func (m *mockTPMValidator) validate(
+	_ context.Context, _ *slog.Logger, params tpm.ValidateParams,
+) (*tpm.ValidatedTPM, error) {
+	m.lastCalledParams = params
+
+	solution, err := params.Solve(&attest.EncryptedCredential{
+		Secret:     []byte("mock-secret"),
+		Credential: []byte("mock-credential"),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if !bytes.Equal(solution, []byte("mock-solution")) {
+		return nil, trace.AccessDenied("invalid solution")
+	}
+
+	return m.returnValidatedTPM, m.returnErr
+}
+
+func TestServer_RegisterUsingTPMMethod(t *testing.T) {
+	ctx := context.Background()
+	mockValidator := &mockTPMValidator{}
+	p, err := newTestPack(ctx, t.TempDir(), func(server *Server) error {
+		server.tpmValidator = mockValidator.validate
+		return nil
+	})
+	require.NoError(t, err)
+	auth := p.a
+
+	sshPrivateKey, sshPublicKey, err := testauthority.New().GenerateKeyPair()
+	require.NoError(t, err)
+	tlsPublicKey, err := PrivateKeyToPublicKeyTLS(sshPrivateKey)
+	require.NoError(t, err)
+
+	attParams := &proto.TPMAttestationParameters{
+		Public: []byte("mock-public"),
+	}
+
+	tokenSpec := func(mutate func(v2 *types.ProvisionTokenSpecV2)) types.ProvisionTokenSpecV2 {
+		spec := types.ProvisionTokenSpecV2{
+			JoinMethod: types.JoinMethodTPM,
+			Roles:      []types.SystemRole{types.RoleNode},
+			TPM: &types.ProvisionTokenSpecV2TPM{
+				Allow: []*types.ProvisionTokenSpecV2TPM_Rule{
+					{
+						Description:  "ekpub only",
+						EKPublicHash: "mock-ekpub-hashed",
+					},
+					{
+						Description:         "ekcert only",
+						EKCertificateSerial: "mock-ekcert-serial",
+					},
+					{
+						Description:         "both",
+						EKPublicHash:        "mock-ekpub-hashed-1",
+						EKCertificateSerial: "mock-ekcert-serial-1",
+					},
+				},
+			},
+		}
+		if mutate != nil {
+			mutate(&spec)
+		}
+		return spec
+	}
+	joinRequest := func() *types.RegisterUsingTokenRequest {
+		return &types.RegisterUsingTokenRequest{
+			HostID:       "host-id",
+			Role:         types.RoleNode,
+			PublicTLSKey: tlsPublicKey,
+			PublicSSHKey: sshPublicKey,
+		}
+	}
+
+	caPool := x509.NewCertPool()
+	require.True(t, caPool.AppendCertsFromPEM([]byte(apifixtures.TLSCACertPEM)))
+
+	allowRulesNotMatched := require.ErrorAssertionFunc(func(t require.TestingT, err error, i ...interface{}) {
+		require.ErrorContains(t, err, "validated tpm attributes did not match any allow rules")
+		require.True(t, trace.IsAccessDenied(err))
+	})
+	tests := []struct {
+		name          string
+		setEnterprise bool
```

Comment:
Suggestion: invert, so it's only set in the one test where it is relevant?

```suggestion
		setOSS bool
```

Reply:
7bb297d836